### DISCUSSION TITLE
Help indexing and calling indicators

### DISCUSSION THREAD

**Discussion by eposada10** [kernc/backtesting.py#965] [Q&A]
Hi im currently working on a project in which I want to integrate the backtesting module so that the user is able to define the indicators via a GUI. For this reason, indicators cannot be created beforehand, but must be created when receiving the user input. My first idea was declaring the indicators using a for loop and indexing them as lists, this didnt work.

 As the self.I() function returns a ndarray i tryed indexing multiple ndarrays that contain the calculatios of the indicator using TA-lib.

However when i run the backtest i dont get any errors but it fails to follow the strategy. 

i used the example given in the main page and modified it as follow:

` 

class SmaCross(Strategy):
    
    n1 = 10
    n2 = 20
    
    def init(self):

        self.sma1 = self.I(ta.SMA, self.data.Close, self.n1)
        self.sma2 = self.I(ta.SMA, self.data.Close, self.n2)
        self.ind = np.empty([2, len(self.data)])  # create an empty 2D array
        self.ind[0] = self.sma1  # assign sma1 to the first row of the array
        self.ind[1] = self.sma2  # assign sma2 to the second row of the array

    def next(self):
        
        if crossover(self.ind[0], self.ind[1]):
            self.position.close()
            self.buy()

        elif crossover(self.ind[1], self.ind[0]):
            self.position.close()
            self.sell()

bt = Backtest(GOOG, SmaCross, cash=100_000, commission=.002)
stats = bt.run()
stats

 `

even thou im calculating exactly the same indicators as in the example, it seems like saving them into an array and then calling them generates this behaviour in which the strategy doesnt execute at all

The results i get are the following.

![image](https://user-images.githubusercontent.com/131006646/232524032-11dbed50-2a9e-4aa1-ba6e-c41848f2ea7e.png)

Any help on how to fix this would be greatly appreciated. 

Thank you

**Comment by kernc** [comment]
The backtest learns about declared indicators in this way:
https://github.com/kernc/backtesting.py/blob/0ce24d80b1bcb8120d95d31dc3bb351b1052a27d/backtesting/backtesting.py#L1191-L1194
Obviously, `isinstance(self.ind, _Indicator)` check fails when `self.ind` is a ndarray ...
